kernel/src/hil/public_key_crypto/signature.rs

```diff
@@ -0,0 +1,59 @@
+// Licensed under the Apache License, Version 2.0 or the MIT License.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// Copyright Tock Contributors 2024.
+
+//! Interface for verifying signatures.
+
+use crate::ErrorCode;
+
+/// This trait provides callbacks for when the verification has completed.
+pub trait ClientVerify<const HL: usize, const SL: usize> {
+    /// Called when the verification is complete.
+    ///
+    /// If the verification operation did not encounter any errors, `result`
+    /// will be set to `Ok()`. If the signature was correctly verified `result`
+    /// will be `Ok(true)`. If the signature did not match the hash `result`
+    /// will be `Ok(false)`.
+    ///
+    /// If verification operation did encounter errors `result` will be `Err()`
+    /// with an appropriate `ErrorCode`. Valid `ErrorCode`s include:
+    ///
+    /// - `CANCEL`: the operation was cancelled.
+    /// - `FAIL`: an internal failure.
+    fn verification_done(
+        &self,
+        result: Result<bool, ErrorCode>,
+        hash: &'static mut [u8; HL],
+        signature: &'static mut [u8; SL],
+    );
+}
+
+/// Verify a signature.
+///
+/// This is a generic interface, and it is up to the implementation as to the
+/// signature verification algorithm being used.
+///
+/// - `HL`: The length in bytes of the hash.
+/// - `SL`: The length in bytes of the signature.
+pub trait SignatureVerify<'a, const HL: usize, const SL: usize> {
+    /// Set the client instance which will receive the `verification_done()`
+    /// callback.
+    fn set_verify_client(&self, client: &'a dyn ClientVerify<HL, SL>);
+
+    /// Verify the signature matches the given hash.
+    ///
+    /// If this returns `Ok(())`, then the `verification_done()` callback will
+    /// be called. If this returns `Err()`, no callback will be called.
+    ///
+    /// The valid `ErrorCode`s that can occur are:
+    ///
+    /// - `OFF`: the underlying digest engine is powered down and cannot be
+    ///   used.
+    /// - `BUSY`: there is an outstanding operation already in process, and the
+    ///   verification engine cannot accept another request.
+    fn verify(
```

Comment:
I am a little confused by this interface. What is the hash generated with? Usually the signature will specify the digest algorithm. It looks like we pass a particular hasher in, but then there's no way to check if it's the correct one for the signature? Is the intention that the higher layer has a bunch of different verify references, and invokes the right one based on the embedded digest?

Reply:
What about something like this? If `hil::digest` specifies `Hash` types, then `SignatureVerify` can template over that (and not just the length). An implementation can then specify which implementation of `Hash` it supports.

The cost here is that I think we have to update `hil::digest` to match (ie use `Hash` types and not just const lengths). Thoughts?

Reply:
I've been thinking about it over the past week and came up with the same conclusion. If it's typed based on the hash algorithm (including length) as well as the public key algorithm (rather than lengths), then all should be good.